codes/zig/chapter_array_and_linkedlist/array.zig

```diff
@@ -0,0 +1,145 @@
+// File: array.zig
+// Created Time: 2023-01-07
+// Author: sjinzh (sjinzh@gmail.com)
+
+const std = @import("std");
+const inc = @import("include");
+
+// 随机返回一个数组元素
+pub fn randomAccess(nums: []i32) i32 {
+    // 在区间 [0, nums.len) 中随机抽取一个整数
+    var randomIndex = std.crypto.random.intRangeLessThan(usize, 0, nums.len);
+    // 获取并返回随机元素
+    var randomNum = nums[randomIndex];
+    return randomNum;
+}
+
+// 扩展数组长度（运行时方法）
+pub fn extend(mem_allocator: std.mem.Allocator, nums: []i32, enlarge: usize) ![]i32 {
+    // 初始化一个扩展长度后的数组
+    var res = try mem_allocator.alloc(i32, nums.len + enlarge);
+    std.mem.set(i32, res, 0);
+    // 将原数组中的所有元素复制到新数组
+    std.mem.copy(i32, res, nums);
+    // 返回扩展后的新数组
+    return res;
+}
+
+// 扩展数组长度（编译期方法A：初始化新数组并进行元素拷贝）
+pub fn extendComptimeA(comptime nums: anytype, comptime enlarge: i32) [nums.len + enlarge]i32 {
+    // 初始化一个扩展长度后的数组
+    var res = [_]i32{0} ** (nums.len + enlarge);
```

Comment:
Is the length of this array determined in the compilation? If so, why could we use the non-constant number to set its length?

Reply:
> Is the length of this array determined in the compilation? If so, why could we use the non-constant number to set its length?

It's a good question! 
"In Zig, the programmer can label variables as comptime. This guarantees to the compiler that every load and store of the variable is performed at compile-time. ". For more details, please see Zig Language Reference(https://ziglang.org/documentation/0.10.0/).

Reply:
Good explanation! More 2 questions:

1. Is `var res = [_]i32{0} ** (nums.len + enlarge);` this array be stored on stack or heap?
2. It is worth discussing the logical correctness to extend array size during compile-time. "comptime variables" means that "we already know the array should be extended", thus why don't we directly create an array with sufficient size?

Reply:
> Good explanation! More 2 questions:
> 
> 1. Is `var res = [_]i32{0} ** (nums.len + enlarge);` this array be stored on stack or heap?
> 2. It is worth discussing the logical correctness to extend array size during compile-time. "comptime variables" means that "we already know the array should be extended", thus why don't we directly create an array with sufficient size?

1. Refer to https://ziglang.org/documentation/0.10.0/#Where-are-the-bytes, "Just like string literals, const declarations, when the value is known at [comptime](https://ziglang.org/documentation/0.10.0/#comptime), are stored in the global constant data section. Also [Compile Time Variables](https://ziglang.org/documentation/0.10.0/#Compile-Time-Variables) are stored in the global constant data section.". So i think the right part "[_]i32{0} ** (nums.len + enlarge)" is in the global constant data section,  because "comptime nums" and "comptime enlarge" , and the left part "res" is in stack data section.
2. Yes, we can directly create an array with sufficient size if we know the array should be extended. Furthermore, if we already know the size of expanded array, we can specify the exact size in comptime for memory saving especially running in the embedded devices.

Reply:
I see. I recommend deleting it. As far as we discussed, the "extending size for array" is a concept only meaningful for the runtime. In other words, we will never extend the array size in compile-time.

Reply:
> I see. I recommend deleting it. As far as we discussed, the "extending size for array" is a concept only meaningful for the runtime. In other words, we will never extend the array size in compile-time.

done.